rust/parser/src/state_machine.rs

```diff
@@ -0,0 +1,647 @@
+use alloc::{collections::BTreeMap, string::String, vec::Vec};
+
+use scale_decode::Visitor;
+
+use merkleized_metadata::{
+    types::{Type, TypeDef, TypeRef},
+    ExtraInfo, TypeResolver,
+};
+
+use crate::{
+    decoding_commons::OutputCard,
+    state::{State, StateError, StateInputCompound, StateInputCompoundItem, StateOutput},
+};
+
+/// Implements the state machine responsible for decoding SCALE-encoded data
+/// into a human-readable format, represented as a list of output cards.
+///
+/// On one hand, this state machine follows the visitor protocol, handling
+/// decoding logic for each supported type. On the other hand, it delegates
+/// to the current state the decision-making regarding the generated output cards and
+/// transitions to the next state.
+///
+/// For more information on how state transitions and output generation are handled,
+/// refer to the [`State`] trait documentation.
+pub struct TypeRegistry(BTreeMap<u32, Vec<Type>>);
+
+impl TypeRegistry {
+    pub fn new<'a>(types: impl Iterator<Item = &'a Type>) -> Self {
+        let tree: BTreeMap<u32, Vec<Type>> = types.fold(Default::default(), |mut map, ty| {
```

Comment:
Is there a specific case when there may be multiple types corresponding to a single `type_id`? Is that how we store enum variants?

Reply:
Yes, a list of enum variants is stored per enum type id